pkg/controller/admissionchecks/multikueue/remoteclient.go

```diff
@@ -0,0 +1,183 @@
+	watchCtx, watchCancel := context.WithCancel(watchCtx)
+	ret := &remoteController{
+		localClient: localClient,
+		watchCtx:    watchCtx,
+		watchCancel: watchCancel,
+		wlUpdateCh:  wlUpdateCh,
+	}
+	return ret
+}
+
+func (rc *remoteController) UpdateConfig(kubeConfigs map[string][]byte) error {
+	if rc.remoteClients == nil {
+		rc.remoteClients = make(map[string]*remoteClient, len(kubeConfigs))
+	}
+
+	for clusterName, c := range rc.remoteClients {
+		if kubeconfig, found := kubeConfigs[clusterName]; found {
+			if err := c.setConfig(kubeconfig); err != nil {
+				delete(rc.remoteClients, clusterName)
+				return err
+			}
+		} else {
+			c.watchCancel()
+			delete(rc.remoteClients, clusterName)
+		}
+	}
+	// create the missing ones
+	for clusterName, kubeconfig := range kubeConfigs {
+		if _, found := rc.remoteClients[clusterName]; !found {
+			c := newRemoteClient(rc.watchCtx, rc.localClient, rc.wlUpdateCh)
+
+			c.restConfigFromKubeConfigOverride = rc.restConfigFromKubeConfigOverride
+			c.newWithWatchOverride = rc.newWithWatchOverride
+
+			if err := c.setConfig(kubeconfig); err != nil {
+				return err
+			}
+			rc.remoteClients[clusterName] = c
+		}
+	}
+
+	return nil
+}
+
+func (cc *remoteController) IsActive() bool {
+	return cc != nil && len(cc.remoteClients) > 0
+}
+
+type remoteClient struct {
+	localClient  client.Client
+	client       client.WithWatch
+	wlUpdateCh   chan<- event.GenericEvent
+	rootWatchCtx context.Context
+	watchCancel  context.CancelFunc
+	watchItf     watch.Interface
+	kubeconfig   []byte
+
+	// For testing only.
+	restConfigFromKubeConfigOverride func([]byte) (*rest.Config, error)
+	newWithWatchOverride             func(config *rest.Config, options client.Options) (client.WithWatch, error)
+}
+
+func newRemoteClient(watchCtx context.Context, localClient client.Client, wlUpdateCh chan<- event.GenericEvent) *remoteClient {
+	rc := &remoteClient{
+		wlUpdateCh:   wlUpdateCh,
+		localClient:  localClient,
+		rootWatchCtx: watchCtx,
+	}
+
+	return rc
+}
+
+// setConfig - will try to recreate the k8s client and restart watching if the new config is different than
+// the one currently used.
+func (rc *remoteClient) setConfig(kubeconfig []byte) error {
+	if equality.Semantic.DeepEqual(kubeconfig, rc.kubeconfig) {
+		return nil
+	}
+
+	if rc.watchCancel != nil {
+		rc.watchCancel()
+		rc.watchCancel = nil
+	}
+
+	var err error
+	var cfg *rest.Config
+
+	if rc.restConfigFromKubeConfigOverride != nil {
+		cfg, err = rc.restConfigFromKubeConfigOverride(kubeconfig)
+	} else {
+		cfg, err = clientcmd.RESTConfigFromKubeConfig(kubeconfig)
+	}
+	if err != nil {
+		return err
+	}
+
+	var remoteClient client.WithWatch
+	if rc.newWithWatchOverride != nil {
+		remoteClient, err = rc.newWithWatchOverride(cfg, client.Options{Scheme: rc.localClient.Scheme()})
+	} else {
+		remoteClient, err = client.NewWithWatch(cfg, client.Options{Scheme: rc.localClient.Scheme()})
+	}
```

Comment:
It's worth having a single override that handles both steps: obtaining the configuration and creating a client.